### PR TITLE
Gh codespace

### LINKED ISSUE #662: Add .devcontainers
Same than:
- https://github.com/jhipster/generator-jhipster/issues/17612
- https://github.com/jhipster/generator-jhipster/pull/17792

The dev to do here:
- add a new business context in `generator/setup/codespaces`
- the goal is to generate 2 new files, in the project:
  - `.devcontainer/devcontainer.json`
  - `.devcontainer/Dockerfile` -> must be Java 17, instead of Java 11
- the API could be: `/api/setup/codespaces`

### PR DESCRIPTION
**Pull request by matthieulapatate**
fix #662

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
@matthieulapatate : you commit with root user, so wrong config :)

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/Codespace.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package tech.jhipster.lite.generator.setup.codespace.domain;
+
+public class Codespace {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/Codespace.java:3]: useless class, can be deleted

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainService.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package tech.jhipster.lite.generator.setup.codespace.domain;
+
+import static tech.jhipster.lite.generator.project.domain.Constants.MAIN_RESOURCES;
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainService.java:3]: not used, can be removed

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationService.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package tech.jhipster.lite.generator.setup.codespace.application;
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationService.java:1]: change location from codespace -> codespaces

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationService.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import org.springframework.stereotype.Service;
+import tech.jhipster.lite.generator.project.domain.Project;
+import tech.jhipster.lite.generator.setup.codespace.domain.CodespaceService;
+
+@Service
+public class CodespaceApplicationService {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationService.java:8]: CodespaceApplicationService -> CodespacesApplicationService

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainService.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  public void init(Project project) {
+    addJSON(project);
+    addDocker(project);
+  }
+
+  public void addJSON(Project project) {
+    projectRepository.add(project, SOURCE, "devcontainer.json", DEVCONTAINER_DEST, "devcontainer.json");
+  }
+
+  public void addDocker(Project project) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainService.java:29]:
```suggestion
  public void addDockerfile(Project project) {
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainService.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void init(Project project) {
+    addJSON(project);
+    addDocker(project);
+  }
+
+  public void addJSON(Project project) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainService.java:25]:
```suggestion
  public void addConfig(Project project) {
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceService.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package tech.jhipster.lite.generator.setup.codespace.domain;
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceService.java:1]:
```suggestion
package tech.jhipster.lite.generator.setup.codespaces.domain;
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceService.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+
+import tech.jhipster.lite.generator.project.domain.Project;
+
+public interface CodespaceService {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceService.java:5]:
```suggestion
public interface CodespacesService {
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/config/CodespaceBeanConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package tech.jhipster.lite.generator.setup.codespace.infrastructure.config;
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/config/CodespaceBeanConfiguration.java:1]:
```suggestion
package tech.jhipster.lite.generator.setup.codespaces.infrastructure.config;
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResource.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+@Tag(name = "Setup")
+class CodespaceResource {
+
+  private final CodespaceApplicationService codespaceApplicationService;
+
+  public CodespaceResource(CodespaceApplicationService codespaceApplicationService) {
+    this.codespaceApplicationService = codespaceApplicationService;
+  }
+
+  @Operation(summary = "Init", description = "Init Maven project with pom.xml and wrapper")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResource.java:26]: the summary is incorrect

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResource.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  @ApiResponse(responseCode = "500", description = "An error occurred while initializing Maven project")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResource.java:27]: the description is incorrect

File: src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResource.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  @PostMapping("/codespaces")
+  @GeneratorStep(id = "codespace")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResource.java:29]:
```suggestion
  @GeneratorStep(id = "codespaces")
```

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationServiceIT.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import static tech.jhipster.lite.generator.setup.codespace.application.CodespaceAssertFiles.assertFilesContainerJson;
+import static tech.jhipster.lite.generator.setup.codespace.application.CodespaceAssertFiles.assertFilesDockerfile;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import tech.jhipster.lite.IntegrationTest;
+import tech.jhipster.lite.generator.project.domain.Project;
+
+@IntegrationTest
+public class CodespaceApplicationServiceIT {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationServiceIT.java:14]: no need public

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceAssertFiles.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
+package tech.jhipster.lite.generator.setup.codespace.application;
+
+import static tech.jhipster.lite.TestUtils.assertFileExist;
+
+import tech.jhipster.lite.generator.project.domain.Project;
+
+public class CodespaceAssertFiles {
+
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceAssertFiles.java:8]: missing private constructor

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationServiceIT.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
+package tech.jhipster.lite.generator.setup.codespace.application;
+
+import static tech.jhipster.lite.TestUtils.assertFileContent;
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/application/CodespaceApplicationServiceIT.java:3]: not used

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import tech.jhipster.lite.UnitTest;
+import tech.jhipster.lite.generator.project.domain.ProjectRepository;
+
+@UnitTest
+@ExtendWith(MockitoExtension.class)
+public class CodespaceDomainServiceTest {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainServiceTest.java:12]: empty class ? no test ?

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceTest.java [hunk 17]
```diff
@@ -0,0 +1,9 @@
+package tech.jhipster.lite.generator.setup.codespace.domain;
+
+import tech.jhipster.lite.UnitTest;
+
+@UnitTest
+public class CodespaceTest {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceTest.java:6]: useless, remove

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/config/CodespaceBeanConfigurationIT.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import tech.jhipster.lite.IntegrationTest;
+import tech.jhipster.lite.generator.setup.codespace.domain.CodespaceDomainService;
+
+@IntegrationTest
+public class CodespaceBeanConfigurationIT {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/config/CodespaceBeanConfigurationIT.java:12]: remove public

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResourceIT.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import tech.jhipster.lite.IntegrationTest;
+import tech.jhipster.lite.generator.project.domain.Project;
+import tech.jhipster.lite.generator.project.infrastructure.primary.dto.ProjectDTO;
+
+@IntegrationTest
+@AutoConfigureMockMvc
+public class CodespaceResourceIT {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespace/infrastructure/primary/rest/CodespaceResourceIT.java:20]: remove public

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Comment by pascalgrimaud** [comment]
More generally, plz rename codespace -> codespaces
See https://github.com/features/codespaces

File: src/test/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainServiceTest.java (hunk 16 above)
Comment (matthieulapatate) [reply src/test/java/tech/jhipster/lite/generator/setup/codespace/domain/CodespaceDomainServiceTest.java:12]: Strange that coverage tests didn't detected that miss.

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/domain/CodespacesDomainService.java [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void init(Project project) {
+    addJSON(project);
+    addDocker(project);
+  }
+
+  public void addJSON(Project project) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/domain/CodespacesDomainService.java:23]:
```suggestion
  public void addConfig(Project project) {
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/domain/CodespacesDomainService.java [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+    projectRepository.add(project, SOURCE, "devcontainer.json", DEVCONTAINER_DEST, "devcontainer.json");
+  }
+
+  public void addDocker(Project project) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/domain/CodespacesDomainService.java:27]:
```suggestion
  public void addDockerfile(Project project) {
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+@Tag(name = "Setup")
+class CodespacesResource {
+
+  private final CodespacesApplicationService codespacesApplicationService;
+
+  public CodespacesResource(CodespacesApplicationService codespacesApplicationService) {
+    this.codespacesApplicationService = codespacesApplicationService;
+  }
+
+  @Operation(summary = "Github Codespaces", description = "Init github codespace configuration files")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java:26]:
```suggestion
  @Operation(summary = "GitHub Codespaces", description = "Init GitHub Codespaces configuration files")
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+  @ApiResponse(responseCode = "500", description = "An error occurred while initializing Github Codespace files.")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java:27]:
```suggestion
  @ApiResponse(responseCode = "500", description = "An error occurred while initializing GitHub Codespace files.")
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+  @PostMapping("/codespaces")
+  @GeneratorStep(id = "github codespaces")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java:29]:
```suggestion
  @GeneratorStep(id = "github-codespaces")
```

File: src/test/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesApplicationServiceIT.java [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+import static tech.jhipster.lite.generator.setup.codespaces.application.CodespacesAssertFiles.assertFilesContainerJson;
+import static tech.jhipster.lite.generator.setup.codespaces.application.CodespacesAssertFiles.assertFilesDockerfile;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import tech.jhipster.lite.IntegrationTest;
+import tech.jhipster.lite.generator.project.domain.Project;
+
+@IntegrationTest
+public class CodespacesApplicationServiceIT {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesApplicationServiceIT.java:13]:
```suggestion
class CodespacesApplicationServiceIT {
```

File: src/test/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesAssertFiles.java [hunk 26]
```diff
@@ -0,0 +1,16 @@
+package tech.jhipster.lite.generator.setup.codespaces.application;
+
+import static tech.jhipster.lite.TestUtils.assertFileExist;
+
+import tech.jhipster.lite.generator.project.domain.Project;
+
+public class CodespacesAssertFiles {
+
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesAssertFiles.java:8]: missing private constructor

File: src/test/java/tech/jhipster/lite/generator/setup/codespaces/domain/CodespacesDomainServiceTest.java [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import tech.jhipster.lite.UnitTest;
+import tech.jhipster.lite.generator.project.domain.Project;
+import tech.jhipster.lite.generator.project.domain.ProjectRepository;
+
+@UnitTest
+@ExtendWith(MockitoExtension.class)
+public class CodespacesDomainServiceTest {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/setup/codespaces/domain/CodespacesDomainServiceTest.java:19]:
```suggestion
class CodespacesDomainServiceTest {
```

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesApplicationService.java [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.stereotype.Service;
+import tech.jhipster.lite.generator.project.domain.Project;
+import tech.jhipster.lite.generator.setup.codespaces.domain.CodespacesService;
+
+@Service
+public class CodespacesApplicationService {
+
+  private final CodespacesService codespacesService;
+
+  public CodespacesApplicationService(CodespacesService codespaceService) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesApplicationService.java:12]:
```suggestion
  public CodespacesApplicationService(CodespacesService codespacesService) {
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesApplicationService.java [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+    this.codespacesService = codespaceService;
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/application/CodespacesApplicationService.java:13]:
```suggestion
    this.codespacesService = codespaceServices;
```

File: src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+class CodespacesResource {
+
+  private final CodespacesApplicationService codespacesApplicationService;
+
+  public CodespacesResource(CodespacesApplicationService codespacesApplicationService) {
+    this.codespacesApplicationService = codespacesApplicationService;
+  }
+
+  @Operation(summary = "GitHub Codespaces", description = "Init GitHub Codespaces configuration files")
+  @ApiResponse(responseCode = "500", description = "An error occurred while initializing GitHub Codespace files.")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/setup/codespaces/infrastructure/primary/rest/CodespacesResource.java:27]:
```suggestion
  @ApiResponse(responseCode = "500", description = "An error occurred while initializing GitHub Codespaces files.")
```

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Comment by pascalgrimaud** [comment]
I fixed the last items and committed directly to your branch.
Then, I managed to test the code:
- generate a project
- add Codespaces
- go into .devcontainer folder
- build the image: `docker build -t jhipster/codespaces . --build-arg NODE_VERSION='lts/*'`
- start the container: `docker run --rm -it jhipster/codespaces bash`
- then, I manage to check the different version of tools:

image

**Review by pascalgrimaud — APPROVED** [review]

**Pull request merged**